### PR TITLE
HPCC-28562 - Add compiler flags to metric sinks to improve security

### PR DESCRIPTION
**Pull request by kenrowland**
Added flags to improve security

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by rpastrana — COMMENTED** [review]
@kenrowland the changes seem fine, but there's no explicit indication how the flags improve security.  
Is there anything particular to this project that prompted this particular change? Should this be applied to other projects?
Are the security improvements only applicable to certain compilers? 
Is it ok to ignore other compilers?

File: system/metrics/sinks/file/fileSink.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ void FileMetricSink::writeMeasurementToFile(const std::shared_ptr<IMetric> &pMet
         size_t countBucketValues = values.size();
         __uint64 cumulative = 0;
 
-        for (int i=0; i < countBucketValues - 1; ++i)
+        for (size_t i=0; i < countBucketValues - 1; ++i)
```
Comment (rpastrana) [inline system/metrics/sinks/file/fileSink.cpp:84]: no mention of this type of change in commit message.

Comment (kenrowland) [reply system/metrics/sinks/file/fileSink.cpp:84]: The above line gave an error/warning with the new flags.

**Comment by kenrowland** [comment]
The flags used are the same that Gavin gave to me (including the compiler type) when updating the build options for HIDL.

File: system/metrics/sinks/file/fileSink.cpp (hunk 1 above)
Comment (rpastrana) [reply system/metrics/sinks/file/fileSink.cpp:84]: Commit message should mention this

**Comment by rpastrana** [comment]
> @rpastrana The flags used are the same that Gavin gave to me (including the compiler type) when updating the build options for HIDL.

@kenrowland the reason for the flags should be documented either inline comment, or in commit message. The question about other compilers stands.

**Comment by rpastrana** [comment]
@ghalliday please take a look

**Review by ghalliday — APPROVED** [review]

**Comment by ghalliday** [comment]
I think the PR is ok as it stands.  Better would be to extract the strict options into a shared symbol, and then use that elsewhere.  E.g. something like:

```
if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_CLANG)
    SET (STRICT_CXX_FLAGS "-Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor")
else()
    SET (STRICT_CXX_FLAGS "")
endif()

SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${STRICT_CXX_FLAGS}")
```

**Comment by kenrowland** [comment]
Added definition in main cmake file for use by any component. Sorry for the delay in requesting a review, but I wanted to make sure all smoketest builds completed since this is change to a common build file used by all components. 

Also changed the HIDL compiler to use new define. 

Once approved, I will rebase

**Review by ghalliday — APPROVED** [review]
Looks good.  There are several other places which could now be changed to use this new symbol:

Search for

-Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor

in the cmake files in following projects: (deftype, eclagent, hql, hqlcpp, hthor, jlib)

**Comment by kenrowland** [comment]
@ghalliday or @richardkchapman Please merge

**Review by ghalliday — APPROVED** [review]

**Pull request merged**